### PR TITLE
Withdraw liquidity functionality

### PR DESCRIPTION
**Pull request by KarishmaBothara**
Please review the logic... I am able to withdraw liquidity

### CODE REVIEW

File: src/redux/epics/liquidity/liquidityRate.epic.ts [hunk 1]
```diff
@@ -22,24 +26,41 @@ export const getCoreLiquidityPriceEpic = (
...
-                    .mul(coreAssetReserve)
-                    .div(tradeAssetReserve)
-                    .addn(1);
+            if (liquidityAction === ADD_LIQUIDITY) {
+                let coreAmount: Amount | BN = assetAmount;
+                if (!coreAssetReserve.isZero() || !tradeAssetReserve.isZero()) {
+                    coreAmount = assetAmount
+                        .mul(coreAssetReserve)
+                        .div(tradeAssetReserve)
+                        .isubn(1);
```
Comment (KarishmaBothara) [inline src/redux/epics/liquidity/liquidityRate.epic.ts:46]: add isubn(1) here... cause with .addn(1) the answer doesn't seem correct...

Comment (holygits) [reply src/redux/epics/liquidity/liquidityRate.epic.ts:46]: could you add some tests for the edge cases that this solves

Comment (KarishmaBothara) [reply src/redux/epics/liquidity/liquidityRate.epic.ts:46]: ya..

Comment (KarishmaBothara) [reply src/redux/epics/liquidity/liquidityRate.epic.ts]: done..

**Comment by KarishmaBothara** [comment]
To test with this set up, need to update assetInfo for CENNZ and CPAY
Screen Shot 2021-04-06 at 1 48 28 PM
Screen Shot 2021-04-06 at 1 48 53 PM

File: src/components/TxSummary/BufferDialog/BufferAsWholeNumber.tsx [hunk 2]
```diff
@@ -1,41 +1,41 @@
-import React, {FC} from 'react';
```
Comment (KarishmaBothara) [inline src/components/TxSummary/BufferDialog/BufferAsWholeNumber.tsx:1]: This file is not used anywhere... But once we have one more hand testing all the functionality, and it works fine.. I will remove this file..

File: src/components/TxSummary/TxSummary.tsx [hunk 3]
```diff
@@ -1,165 +1,165 @@
-import BN from 'bn.js';
```
Comment (KarishmaBothara) [inline src/components/TxSummary/TxSummary.tsx:1]: same..This file is not used anywhere

File: src/components/TxSummary/TxSummaryBodyForBuy.tsx [hunk 4]
```diff
@@ -1,46 +1,46 @@
-import React, {FC} from 'react';
```
Comment (KarishmaBothara) [inline src/components/TxSummary/TxSummaryBodyForBuy.tsx:1]: same..This file is not used anywhere

File: src/components/TxSummary/TxSummaryBodyForWith.tsx [hunk 5]
```diff
@@ -1,33 +1,33 @@
-import React, {FC} from 'react';
```
Comment (KarishmaBothara) [inline src/components/TxSummary/TxSummaryBodyForWith.tsx:1]: same..This file is not used anywhere

File: src/components/TxSummary/TxSummaryEstimatedTxFeeForBody.tsx [hunk 6]
```diff
@@ -1,56 +1,56 @@
-import BN from 'bn.js';
```
Comment (KarishmaBothara) [inline src/components/TxSummary/TxSummaryEstimatedTxFeeForBody.tsx:1]: same..This file is not used anywhere

**Review by holygits — COMMENTED** [review]
initial review.
couple more files and still tests to check
I won't check deployment stuff here

File: src/components/AdvancedSetting/index.tsx [hunk 7]
```diff
@@ -136,11 +148,12 @@ const AdvancedSetting: FC<SummaryOrErrorProps> = ({
                         {state.spinner && (
                             <>
                                 <TextInput
+                                    type={'number'}
                                     title={'Slippage'}
                                     placeholder={'Maximum slippage percent'}
-                                    value={buffer}
+                                    value={buffer * 100}
```
[the remark this answers is not in the document]
Comment (KarishmaBothara) [reply src/components/AdvancedSetting/index.tsx:154]: yes

File: src/components/Dialog/TxDialog/TxDialogBody.tsx [hunk 8]
```diff
@@ -40,22 +39,93 @@ const BodyForFinalised: FC<BodyForFinalisedProps> = ({
...
+            case SWAP_INPUT:
+                message = `You successfully exchanged ${toAssetAmount.asString(
+                    toAssetDecimalPlaces
+                )} ${toAssetSymbol} with 
+                    ${fromAssetAmount.asString(fromAssetDecimalPlaces)} ${fromAssetSymbol}.`;
+                break;
+            case ADD_LIQUIDITY:
+                message = `You successfully added ${toAssetAmount.asString(
+                    fromAssetDecimalPlaces
+                )} ${fromAssetSymbol} and
```
[the remark this answers is not in the document]
Comment (KarishmaBothara) [reply src/components/Dialog/TxDialog/TxDialogBody.tsx:68]: ya... its correct... cause the order in which we get is generic for all operations, but liquidity uses it this way... I'll add a comment or see If I can change it..

File: static.config.js [hunk 9]
```diff
@@ -3,8 +3,8 @@ import React from 'react'
 export default {
     entry: 'index.tsx',
     //maxThreads: 1,
-    crossOriginLoading: 'anonymous',
-    productionSourceMaps: true,
+    productionSourceMaps: false,
+    siteRoot: 'https://cennzx.centrality.me',
```
[the remark this answers is not in the document]
Comment (KarishmaBothara) [reply static.config.js:7]: I'll remove this we won't need this..

File: src/components/Dialog/TxDialog/TxDialogBody.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (KarishmaBothara) [reply src/components/Dialog/TxDialog/TxDialogBody.tsx]: updated the code, to make this look more readable...

File: src/components/AssetDropDown.tsx [hunk 10]
```diff
@@ -85,12 +85,23 @@ const Container = styled.div`
...
+            <Container>
+                <Icon src={require(`./../images/${symbol.toLowerCase()}.png`)} />
+                <Text className="symbol">{symbol}</Text>
+            </Container>
+        );
+    } catch (e) {
+        // when a new asset is added and image does not exist
+        return (
+            <Container>
+                <Text className="symbol">{symbol}</Text>
```
[the remark this answers is not in the document]
Comment (KarishmaBothara) [reply src/components/AssetDropDown.tsx:101]: Screen Shot 2021-04-21 at 10 22 28 AM
Screen Shot 2021-04-21 at 10 23 24 AM

File: src/components/Dialog/TxDialog/TxDialogBody.tsx [hunk 11]
```diff
@@ -41,77 +41,41 @@ const BodyForFinalised: FC<BodyForFinalisedProps> = ({
...
+        const coreAssetSymbol = assetInfo[coreAssetId].symbol;
+        const assetDecimalPlaces = assetInfo[assetId].decimalPlaces;
+        const assetSymbol = assetInfo[assetId].symbol;
+        const action = method === ADD_LIQUIDITY ? 'added' : 'withdraw';
+        const place = method === ADD_LIQUIDITY ? 'in' : 'from';
+        if (success) {
+            message = `You successfully ${action} ${assetAmount.asString(assetDecimalPlaces)} ${assetSymbol} and
+                    ${coreAmount.asString(coreAssetDecimalPlaces)} ${coreAssetSymbol} ${place} the pool.`;
+        } else {
+            message = `Your transaction to ${action} liquidity of ${assetAmount.asString(
```
Comment (holygits) [inline src/components/Dialog/TxDialog/TxDialogBody.tsx:72]: nit: this will say 'your transaction to added liquidity'

File: src/components/Dialog/TxDialog/TxDialogBody.tsx [hunk 12]
```diff
@@ -41,77 +41,41 @@ const BodyForFinalised: FC<BodyForFinalisedProps> = ({
...
+    } else if (method === ADD_LIQUIDITY || method === REMOVE_LIQUIDITY) {
+        const [assetId, coreAssetId, coreAmount, assetAmount] = params as AssetSwapParams;
+        const coreAssetDecimalPlaces = assetInfo[coreAssetId].decimalPlaces;
+        const coreAssetSymbol = assetInfo[coreAssetId].symbol;
+        const assetDecimalPlaces = assetInfo[assetId].decimalPlaces;
+        const assetSymbol = assetInfo[assetId].symbol;
+        const action = method === ADD_LIQUIDITY ? 'added' : 'withdraw';
+        const place = method === ADD_LIQUIDITY ? 'in' : 'from';
+        if (success) {
+            message = `You successfully ${action} ${assetAmount.asString(assetDecimalPlaces)} ${assetSymbol} and
```
Comment (holygits) [inline src/components/Dialog/TxDialog/TxDialogBody.tsx:69]: nit: this will say 'You successfully withdraw'

File: src/components/Dialog/TxDialog/TxDialogBody.tsx (hunk 11 above)
Comment (KarishmaBothara) [reply src/components/Dialog/TxDialog/TxDialogBody.tsx:73]: yes... the ${action} will either be added or withdraw depending on if its add liquidity or remove and the message would be exactly as above..

File: src/components/Dialog/TxDialog/TxDialogBody.tsx (hunk 12 above)
Comment (KarishmaBothara) [reply src/components/Dialog/TxDialog/TxDialogBody.tsx:70]: yes... the ${action} will either be added or withdraw depending on if its add liquidity or remove and the message would be exactly as above..

Comment (holygits) [reply src/components/Dialog/TxDialog/TxDialogBody.tsx:70]: I'm pointing out the grammar will be wrong because it's present tense on success and past tense on failure

File: src/components/TextInput.tsx [hunk 13]
```diff
@@ -65,7 +65,7 @@ const TextInput: FC<InputProps> = ({value, onChange, title, placeholder, message
                 </span>
             </Top>
             <Input onChange={onChange}>
-                <input value={value} placeholder={placeholder} />
+                <input type={type} value={value} placeholder={placeholder} />
```
[the remark this answers is not in the document]
Comment (KarishmaBothara) [reply src/components/TextInput.tsx:68]: This component is only used to adjust the slippage... which was suppose to be number... I can change the component name to InputBox which would accept type and make that type of input...

File: src/components/Dialog/TxDialog/TxDialogBody.tsx (hunk 12 above)
Comment (KarishmaBothara) [reply src/components/Dialog/TxDialog/TxDialogBody.tsx:70]: I see.. will update

File: src/components/Dialog/TxDialog/TxDialogBody.tsx (hunk 11 above)
Comment (KarishmaBothara) [reply src/components/Dialog/TxDialog/TxDialogBody.tsx:74]: updated the message..

File: src/pages/exchange/exchange.tsx [hunk 14]
```diff
@@ -149,6 +160,14 @@ export const Exchange: FC<ExchangeProps & ExchangeDispatchProps> = props => {
...
     const assetForEmptyPool = error.find(err => err instanceof EmptyPool);
     const formErrors = state.touched ? getFormErrors(props) : new Map<FormSection, FormErrorTypes[]>();
+    let maxBuy;
+    if (outputReserve && toAssetBalance) {
+        // Take minimum of user and pool balance for max Buy amount.
+        maxBuy = outputReserve.gt(toAssetBalance) ? toAssetBalance : outputReserve;
+    } else if (outputReserve) {
+        // when no user is selected use pool balance as max..
+        maxBuy = outputReserve;
+    }
```
Comment (holygits) [inline src/pages/exchange/exchange.tsx]: better to keep it simple, if user has `0` or unset then max should be `0`
`let maxBuy = Math.min(outputReserve, toAssetBalance);`

Comment (KarishmaBothara) [reply src/pages/exchange/exchange.tsx]: OutputReserve and toAssetBalance both are of type Amount which extends BN...
I tried doing Amount.min(outputReserve, toAssetBalance) but it had some issue.. will try again...

Comment (KarishmaBothara) [reply src/pages/exchange/exchange.tsx]: done

**Review by holygits — APPROVED** [review]
lgtm minus a few nits

File: src/components/Dialog/TxDialog/TxDialog.tsx [hunk 15]
```diff
@@ -60,13 +59,15 @@ export const TxDialog: FC<TxDialogProps> = ({
                     coreAssetId={coreAssetId}
                     error={error}
                     success={success}
-                    recipientAddress={title === 'send' && send.recipientAddress}
+                    recipientAddress={undefined}
+                    // recipientAddress={title === 'send' && send.recipientAddress}
```
[the remark this answers is not in the document]
Comment (KarishmaBothara) [reply src/components/Dialog/TxDialog/TxDialog.tsx]: removed the comment.. have kept recipientAddress field... when we want to support sending to different user we can pass the proper value to it..

File: src/redux/epics/_test_/checkPolkadotExtension.epic.test.ts [hunk 16]
```diff
@@ -1,32 +1,31 @@
...
         name: 'Account 1',
+        meta: {name: 'Account 1'},
         address: '5GDq1kEpNzxWQcnviMRFnp1y8m47kWC1EDEUzgCZQFc4G1Df',
     },
 ];
 accounts$.next(account);
 
-describe('trigger on init, extension detected', () => {
+// FIXME
+describe.skip('trigger on init, polkadot extension detected', () => {
```
[the remark this answers is not in the document]
Comment (KarishmaBothara) [reply src/redux/epics/_test_/checkPolkadotExtension.epic.test.ts:28]: I will create a new issue for this.. the problem with this is, it needs interaction with browser for user to click enable and then it can be tested... need to investigate any possible option to do it..

File: src/redux/epics/_test_/exchange/balance.epic.test.ts [hunk 17]
```diff
@@ -71,3 +72,60 @@ describe('Request asset balance working', () => {
...
+                const state$: Observable<AppState> = new StateObservable(new Subject(), {
+                    ui: {
+                        exchange: {
+                            form: {
+                                fromAsset: 16000,
+                                toAsset: 16001,
+                                feeAssetId: 16001,
+                                fromAssetAmount: new Amount(2),
+                                toAssetAmount: new Amount(1),
+                                signingAccount: '5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty',
```
[the remark this answers is not in the document]
Comment (KarishmaBothara) [reply src/redux/epics/_test_/exchange/balance.epic.test.ts]: done..thanks

File: src/redux/epics/_test_/exchange/price.epic.test.ts [hunk 18]
```diff
@@ -462,11 +617,71 @@ describe('Test when output price epic throws', () => {
...
+    });
+});
+
+describe('Test when output price epic throws', () => {
+    const inputAmount = new Amount('2', AmountUnit.DISPLAY);
+    const triggers = [setToAssetAmount(inputAmount)];
+    triggers.forEach(action => {
+        it(action.type, () => {
+            const testScheduler = new TestScheduler((actual, expected) => {
+                // somehow assert the two objects are equal
```
[the remark this answers is not in the document]
Comment (KarishmaBothara) [reply src/redux/epics/_test_/exchange/price.epic.test.ts]: removed for all the tests...

File: src/redux/reducers/global.reducer.ts [hunk 19]
```diff
@@ -27,6 +39,17 @@ export default handleActions(
...
             draft.metadata = action.payload;
         }),
+        [GlobalActions.ASSET_INFO_UPDATE]: produce((draft: GlobalState, action: UpdateAssetsInfoAction) => {
+            const newAssetList = new Array();
+            const assetList = action.payload;
+            assetList.forEach(asset => {
+                const id = asset[0];
+                const symbol = hexToString(asset[1].symbol);
+                const decimalPlaces = asset[1].decimalPlaces;
+                newAssetList[asset[0]] = {decimalPlaces, symbol, id};
```
[the remark this answers is not in the document]
Comment (KarishmaBothara) [reply src/redux/reducers/global.reducer.ts]: done..
